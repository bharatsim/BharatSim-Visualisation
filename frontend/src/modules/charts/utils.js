import { chartStyleConfig, getColor } from './chartStyleConfig';

function trasformDataForChart(csvData, xColumn, yColumns) {
  return {
    labels: csvData && csvData.data[xColumn],
    datasets:
      csvData &&
      yColumns.map((yColumn, index) => {
        return {
          ...chartStyleConfig,
          label: yColumn,
          borderColor: getColor(index),
          backgroundColor: getColor(index),
          data: csvData && csvData.data[yColumn],
        };
      }),
  };
}

function getYaxisNames(yColumns) {
  return yColumns.map((yColumn) => yColumn.name);
}

export { trasformDataForChart, getYaxisNames };
