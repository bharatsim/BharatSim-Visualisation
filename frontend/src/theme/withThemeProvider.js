import React from 'react';
import { MuiThemeProvider, CssBaseline } from '@material-ui/core';

import theme from './theme';

function withThemeProvider(WrappedComponent) {
  return function WithTheme(props) {
    return (
      <MuiThemeProvider theme={theme}>
        <CssBaseline />
        <WrappedComponent {...props} />
      </MuiThemeProvider>
    );
  };
}

export default withThemeProvider;
