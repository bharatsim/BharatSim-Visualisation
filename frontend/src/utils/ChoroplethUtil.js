import L from 'leaflet';
import chroma from 'chroma-js';

function highlightOnClick(e) {
  const layer = e.target;

  layer.setStyle({
    weight: 2,
    dashArray: '',
    opacity: 1,
    fillOpacity: 0.7,
  });

  if (!L.Browser.ie && !L.Browser.opera && !L.Browser.edge) {
    layer.bringToFront();
  }
}

function resetHighlight(geoJson) {
  return (e) => geoJson.current.resetStyle(e.target);
}

function zoomToFeature(map) {
  return (e) => map.fitBounds(e.target.getBounds());
}

function geoJSONStyle(idDataMap, mapLayerIdName, maxOfMeasure, scale) {
  const colors = Object.keys(scale)
    .sort((v1, v2) => Number(v1) - Number(v2))
    .map((key) => scale[key]);
  const getColor = chroma.scale(colors);
  return (feature) => {
    const gradientPercentage = idDataMap[feature.properties[mapLayerIdName]] / maxOfMeasure;
    return {
      color: '#1f2021',
      weight: 1,
      fillOpacity: 0.5,
      opacity: 0.7,
      fillColor: getColor(gradientPercentage),
    };
  };
}

export { highlightOnClick, resetHighlight, geoJSONStyle, zoomToFeature };
