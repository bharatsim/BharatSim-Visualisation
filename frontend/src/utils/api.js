import { contentTypes, httpMethods } from '../constants/fetch';
import { url as serviceURL } from './url';
import { fetchData, uploadData } from './fetch';

function headerBuilder({ contentType }) {
  return {
    'content-type': contentType,
  };
}

function formDataBuilder(data) {
  const formData = new FormData();
  data.forEach(({ name, value }) => formData.append(name, value));
  return formData;
}

const api = {
  saveDashboard: async ({ widgets, layout, dashboardId, name, count }) => {
    return uploadData({
      url: serviceURL.DASHBOARD_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      data: JSON.stringify({
        dashboardData: { widgets, layout, dashboardId, name, count },
      }),
    });
  },

  addNewDashboard: async ({ name, projectId }) => {
    return uploadData({
      url: serviceURL.INSERT_DASHBOARD,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      data: JSON.stringify({
        dashboardData: { widgets: [], layout: [], name, count: 0, projectId },
      }),
      isCustomErrorHandler: true,
    });
  },

  getAllDashBoardByProjectId: async (projectId) =>
    fetchData({
      url: serviceURL.DASHBOARD_URL,
      query: { projectId, columns: ['name', '_id'] },
    }),

  getAllDashBoard: async () => fetchData({ url: serviceURL.DASHBOARD_URL }),

  uploadFileAndSchema: async ({ file, schema, dashboardId }) =>
    uploadData({
      url: serviceURL.DATA_SOURCES,
      data: formDataBuilder([
        { name: 'datafile', value: file },
        { name: 'schema', value: JSON.stringify(schema) },
        { name: 'dashboardId', value: dashboardId },
      ]),
      headers: headerBuilder({ contentType: contentTypes.FILE }),
    }),

  getCsvHeaders: async (dataSourceId) => fetchData({ url: serviceURL.getHeaderUrl(dataSourceId) }),

  getDatasources: async (dashboardId) =>
    fetchData({
      url: serviceURL.DATA_SOURCES,
      query: { dashboardId },
    }),

  getData: async (datasource, columns) =>
    fetchData({ url: serviceURL.getDataUrl(datasource), query: { columns } }),

  saveProject: async ({ id, ...data }) => {
    const requestObject = {
      url: serviceURL.PROJECT_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      isCustomErrorHandler: true,
    };
    if (id) {
      return uploadData({
        ...requestObject,
        data: JSON.stringify({ projectData: { ...data, id } }),
        method: httpMethods.PUT,
      });
    }
    return uploadData({
      ...requestObject,
      data: JSON.stringify({ projectData: { ...data } }),
      method: httpMethods.POST,
    });
  },

  getProjects: async () => {
    return fetchData({
      url: serviceURL.PROJECT_URL,
    });
  },
  getProject: async (id) => {
    return fetchData({
      url: serviceURL.getProjectUrl(id),
    });
  },
};

export { api };
