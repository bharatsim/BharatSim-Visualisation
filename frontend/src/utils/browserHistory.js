// eslint-disable-next-line import/no-mutable-exports
let BrowserHistory;

function initHistory(history) {
  BrowserHistory = history;
}

function navigateToPreviousPage() {
  BrowserHistory.goBack();
}

function navigateToHome() {
  BrowserHistory.push('/');
}

function reloadPage() {
  // eslint-disable-next-line no-restricted-globals
  location.reload();
}

export { initHistory, navigateToPreviousPage, reloadPage, BrowserHistory, navigateToHome };
