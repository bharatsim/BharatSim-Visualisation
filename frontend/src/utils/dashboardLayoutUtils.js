export function getNewWidgetLayout(numberOfWidgetAdded, cols, count) {
  return createLayout({
    id: `widget-${count}`,
    xPosition: (numberOfWidgetAdded * 6) % cols,
    yPosition: Infinity, // puts it at the bottom
    width: 6,
    height: 2,
  });
}

function createLayout({ id, xPosition, yPosition, width, height }) {
  return {
    i: id,
    x: xPosition,
    y: yPosition,
    w: width,
    h: height,
  };
}
