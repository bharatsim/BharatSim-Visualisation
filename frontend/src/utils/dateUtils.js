const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

function getPeriod(hours) {
  return hours >= 12 ? 'PM' : 'AM';
}

function formatDate(dateString) {
  const date = new Date(dateString);
  if (date.toString() === 'Invalid Date') return '--';
  const DD = date.getDate();
  const MM = date.getMonth();
  const YYYY = date.getFullYear();
  const HH = date.getHours();
  const mm = date.getMinutes();

  return `${DD}-${months[MM]}-${YYYY} at ${HH % 12}:${mm < 10 ? `0${mm}` : mm} ${getPeriod(HH)}`;
}

function currentDate(){
  return new Date();
}

export { formatDate, currentDate };
