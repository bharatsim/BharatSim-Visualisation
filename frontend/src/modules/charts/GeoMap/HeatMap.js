import React, { useEffect, useMemo, useState } from 'react';
import PropTypes from 'prop-types';

import { makeStyles } from '@material-ui/core/styles';
import { MapContainer, ScaleControl, TileLayer } from 'react-leaflet';
import 'leaflet/dist/leaflet.css';
import Box from '@material-ui/core/Box';
import useLoader from '../../../hook/useLoader';
import { api } from '../../../utils/api';
import LoaderOrError from '../../loaderOrError/LoaderOrError';
import { getLatLngCenter, transformDataForHeatMap } from '../../../utils/helper';
import ViewController from '../../../uiComponent/mapLayers/ViewController';
import HeatMapLayer from '../../../uiComponent/mapLayers/HeatMapLayer';
import ColorScaleLegend from '../../../uiComponent/mapLayers/ColorScaleLegend';
import ResizeController from '../../../uiComponent/mapLayers/ResizeController';
import { scale } from '../../../constants/colorScale';
import TimeSlider from '../../../uiComponent/TimeSlider';

const useStyles = makeStyles((theme) => ({
  fullWidthHeight: { height: '100%', width: '100%' },
  mapContainer: {
    height: `calc(100% - ${theme.spacing(15)}px )`,
  },
}));

function HeatMap({ config }) {
  const {
    dataSource,
    geoDimensions: { latitude, longitude },
    geoMetricSeries,
    sliderConfig,
  } = config;
  const { timeMetrics } = sliderConfig;
  const [timeSliderValue, setTimeSliderValue] = useState(2);

  const classes = useStyles();
  const [fetchedData, setFetchedData] = useState();
  const {
    loadingState,
    message,
    startLoader,
    stopLoaderAfterError,
    stopLoaderAfterSuccess,
  } = useLoader();

  useEffect(() => {
    fetchData();
  }, [latitude, longitude, geoMetricSeries, timeMetrics]);

  const locationPoints = useMemo(
    () =>
      transformDataForHeatMap(
        fetchedData,
        latitude,
        longitude,
        geoMetricSeries,
        timeMetrics,
        timeSliderValue,
      ),
    [fetchedData, timeSliderValue],
  );

  const center = getLatLngCenter(locationPoints);

  const maxOfGeoMatrixSeries =
    fetchedData && fetchedData[geoMetricSeries] ? Math.max(...fetchedData[geoMetricSeries]) : 1;

  async function fetchData() {
    const columns = [latitude, longitude, geoMetricSeries];
    if (timeMetrics) {
      columns.push(timeMetrics);
    }
    startLoader();
    return api
      .getData(dataSource, columns)
      .then(({ data }) => {
        stopLoaderAfterSuccess();
        setFetchedData(data);
      })
      .catch(() => {
        stopLoaderAfterError('Unable to fetch data');
      });
  }

  useEffect(() => {
    if (fetchedData && fetchedData[timeMetrics]) {
      setTimeSliderValue(Math.min(...fetchedData[timeMetrics]));
    }
  }, [fetchedData]);

  const onErrorAction = {
    name: 'Retry',
    onClick: fetchData,
  };

  return (
    <div className={timeMetrics ? classes.mapContainer : classes.fullWidthHeight}>
      <LoaderOrError loadingState={loadingState} message={message} errorAction={onErrorAction}>
        <div className={classes.fullWidthHeight} data-testid="map-container">
          <Box px={4}>
            {timeMetrics && fetchedData && fetchedData[timeMetrics] && (
              <Box>
                <TimeSlider
                  defaultValue={Math.min(...fetchedData[timeMetrics])}
                  maxValue={Math.max(...fetchedData[timeMetrics])}
                  minValue={Math.min(...fetchedData[timeMetrics])}
                  step={1}
                  setTimeSliderValue={setTimeSliderValue}
                  title={timeMetrics}
                  timeSliderValue={timeSliderValue}
                />
              </Box>
            )}
          </Box>
          <MapContainer className={classes.fullWidthHeight} center={center} zoom={8} preferCanvas>
            <TileLayer
              attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
              url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
            />
            <HeatMapLayer
              points={locationPoints}
              options={{
                max: maxOfGeoMatrixSeries,
                radius: 25,
                minOpacity: 0.3,
                gradient: scale,
              }}
              key={timeSliderValue}
            />
            <ColorScaleLegend scale={scale} />
            <ResizeController />
            <ViewController center={center} />
            <ScaleControl />
          </MapContainer>
        </div>
      </LoaderOrError>
    </div>
  );
}

HeatMap.propTypes = {
  config: PropTypes.shape({
    dataSource: PropTypes.string.isRequired,
    geoDimensions: PropTypes.shape({ latitude: PropTypes.string, longitude: PropTypes.string })
      .isRequired,
    geoMetricSeries: PropTypes.string.isRequired,
    sliderConfig: PropTypes.shape({ timeMetrics: PropTypes.string }).isRequired,
  }).isRequired,
};

export default HeatMap;
