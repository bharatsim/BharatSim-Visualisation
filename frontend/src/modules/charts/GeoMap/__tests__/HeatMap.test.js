import React, { useEffect } from 'react';
import { fireEvent } from '@testing-library/dom';
import { render } from '@testing-library/react';
import HeatMap from '../HeatMap';
import 'leaflet/dist/leaflet.css';
import { api } from '../../../../utils/api';
import withThemeProvider from '../../../../theme/withThemeProvider';

jest.mock('../../../../utils/api', () => ({
  api: {
    getData: jest.fn().mockResolvedValue({
      data: { lat: [2, 3], long: [1, 2], dataPoints: [2, 3] },
    }),
  },
}));

const mockEffect = useEffect;
jest.mock('react-leaflet', () => ({
  MapContainer: ({ children, whenCreated }) => {
    mockEffect(() => {
      whenCreated({ invalidateSize: jest.fn() });
    }, []);

    return (
      <>
        MapContainer
        {children}
      </>
    );
  },
  TileLayer: () => <>TileLayer</>,
  ScaleControl: () => <>ScaleControl</>,
}));

jest.mock('../../../../uiComponent/mapLayers/HeatMapLayer', () => (props) => (
  <>
    HeatMapLayer
    {JSON.stringify(props)}
  </>
));

jest.mock('../../../../uiComponent/mapLayers/ViewLayer', () => (props) => (
  <>
    ViewAndZoomLayer
    {JSON.stringify(props)}
  </>
));

jest.mock('../../../../uiComponent/mapLayers/ColorScaleLegend', () => (props) => (
  <>
    ColorScaleLegend
    {JSON.stringify(props)}
  </>
));

describe('HeatMap', () => {
  beforeAll(() => {
    jest.useFakeTimers();
  });
  beforeEach(() => {
    jest.runAllTimers();
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  const HeatMapWithProvider = withThemeProvider(HeatMap);

  it('should create a heatmap component', async () => {
    const { container, findByTestId } = render(
      <HeatMapWithProvider
        config={{
          dataSource: 'test.csv',
          geoDimensions: { latitude: 'lat', longitude: 'long' },
          geoMetricSeries: 'dataPoints',
        }}
      />,
    );

    await findByTestId('map-container');

    expect(container).toMatchSnapshot();
  });

  it('should show error if api fail to load', async () => {
    api.getData.mockRejectedValueOnce('error');
    const { findByText, getByText } = render(
      <HeatMapWithProvider
        config={{
          dataSource: 'test.csv',
          geoDimensions: { latitude: 'lat', longitude: 'long' },
          geoMetricSeries: 'dataPoints',
        }}
      />,
    );

    await findByText('Unable to fetch data');

    expect(getByText('Unable to fetch data')).toBeInTheDocument();
  });

  it('should call fetch data on click of retry button', async () => {
    api.getData.mockRejectedValueOnce('error');
    const { findByText, getByText, findByTestId } = render(
      <HeatMapWithProvider
        config={{
          dataSource: 'test.csv',
          geoDimensions: { latitude: 'lat', longitude: 'long' },
          geoMetricSeries: 'dataPoints',
        }}
      />,
    );

    await findByText('Unable to fetch data');

    const retryButton = getByText('Retry');

    fireEvent.click(retryButton);

    await findByTestId('map-container');

    expect(api.getData).toHaveBeenLastCalledWith('test.csv', ['lat', 'long', 'dataPoints']);
  });
});
