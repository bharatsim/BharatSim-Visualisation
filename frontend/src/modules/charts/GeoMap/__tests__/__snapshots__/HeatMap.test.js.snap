// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeatMap should create a heatmap component 1`] = `
<div>
  <div
    class="makeStyles-fullWidthHeight-1"
  >
    <div
      class="makeStyles-fullWidthHeight-1"
      data-testid="map-container"
    >
      MapContainer
      ViewAndZoomLayer
      {"center":[2.5000950751631947,1.4998094835255649]}
      TileLayer
      HeatMapLayer
      {"points":[[2,1,2],[3,2,3]],"options":{"max":3,"radius":75}}
      ScaleControl
    </div>
  </div>
</div>
`;
